File: sprites.py
"""
Game sprite classes: Player, Enemies, Projectiles, Effects
"""
import pygame
import math
import random
from constants import WIDTH, HEIGHT, WHITE, BLACK
from utils import load_game_assets

# Load assets for sprites
ship_img, bullet_img, explosion_img, asteroid_images = load_game_assets()

class MotionTrail(pygame.sprite.Sprite):
    def __init__(self, image, position, alpha=150):
        pygame.sprite.Sprite.__init__(self)
        self.image = image.copy()
        self.image.set_alpha(alpha)
        self.rect = self.image.get_rect()
        self.rect.center = position
        self.lifetime = 15  # frames
        
    def update(self):
        self.lifetime -= 1
        if self.lifetime <= 0:
            self.kill()
        else:
            # Fade out
            alpha = int((self.lifetime / 15) * 150)
            self.image.set_alpha(alpha)

class Sheera(pygame.sprite.Sprite):
    def __init__(self, game_mode="normal"):
        pygame.sprite.Sprite.__init__(self)
        # Load different image based on game mode
        if game_mode == "accelerated":
            try:
                import os
                from constants import assets_dir
                mode_img = pygame.image.load(os.path.join(assets_dir, "ChatGPT Image Jun 12, 2025, 07_17_20 PM.png"))
                scaled_img = pygame.transform.scale(mode_img, (60, 60))
                self.image = pygame.transform.rotate(scaled_img, 90)
            except:
                self.image = ship_img
        elif game_mode == "slowed":
            try:
                import os
                from constants import assets_dir
                mode_img = pygame.image.load(os.path.join(assets_dir, "Generated Image June 12, 2025 - 7_16PM.png"))
                scaled_img = pygame.transform.scale(mode_img, (60, 60))
                self.image = pygame.transform.rotate(scaled_img, 90)
            except:
                self.image = ship_img
        else:
            self.image = ship_img
        self.original_image = self.image
        self.rect = self.image.get_rect()
        self.rect.center = (WIDTH // 2, HEIGHT // 2)
        self.position = pygame.math.Vector2(self.rect.center)
        self.velocity = pygame.math.Vector2(0, 0)
        self.angle = 0
        self.rotation_speed = 1.5
        self.acceleration = 0.2
        self.max_speed = 7
        self.friction = 0.98
        self.shoot_delay = 250
        self.last_shot = pygame.time.get_ticks()
        self.lives = 3
        self.hidden = False
        self.hide_timer = pygame.time.get_ticks()
        self.invulnerable = False
        self.invulnerable_timer = 0
        self.trail_timer = 0
        
        # Glow effect properties
        self.heat = 0
        self.max_heat = 100
        self.heat_increase = 20
        self.heat_cooldown = 1
        self.glow_colors = [
            (255, 255, 0), (255, 165, 0), (255, 69, 0), (255, 0, 0)
        ]
        
        # Shield properties
        self.shield_active = False
        self.shield_strength = 0
        self.max_shield = 100
        self.shield_increase = 2
        self.shield_decrease = 5
        self.shield_colors = [
            (0, 100, 255), (0, 50, 255), (0, 0, 255), (75, 0, 130)
        ]

    def update(self):
        # Handle invulnerability
        if self.invulnerable:
            if pygame.time.get_ticks() - self.invulnerable_timer > 2000:
                self.invulnerable = False
        
        if self.hidden:
            # Check if it's time to respawn (after 3 seconds)
            if pygame.time.get_ticks() - self.hide_timer > 3000:
                self.respawn()
            return
        
        # Handle input
        keys = pygame.key.get_pressed()
        
        # Rotation
        if keys[pygame.K_LEFT]:
            self.angle += self.rotation_speed
        if keys[pygame.K_RIGHT]:
            self.angle -= self.rotation_speed
        
        # Thrust
        if keys[pygame.K_UP]:
            thrust_x = math.cos(math.radians(self.angle)) * self.acceleration
            thrust_y = -math.sin(math.radians(self.angle)) * self.acceleration
            self.velocity.x += thrust_x
            self.velocity.y += thrust_y
            
            # Limit speed
            if self.velocity.length() > self.max_speed:
                self.velocity.scale_to_length(self.max_speed)
        
        # Apply friction
        self.velocity *= self.friction
        
        # Update position
        self.position += self.velocity
        
        # Wrap around screen
        if self.position.x < 0:
            self.position.x = WIDTH
        elif self.position.x > WIDTH:
            self.position.x = 0
        if self.position.y < 0:
            self.position.y = HEIGHT
        elif self.position.y > HEIGHT:
            self.position.y = 0
            
        # Update rect position
        self.rect.center = self.position
        
        # Cool down heat
        self.heat = max(0, self.heat - self.heat_cooldown)
        
        # Rotate ship image
        self.image = pygame.transform.rotate(self.original_image, self.angle)
        self.rect = self.image.get_rect(center=self.rect.center)
    
    def shoot(self):
        now = pygame.time.get_ticks()
        if now - self.last_shot > self.shoot_delay and not self.hidden:
            self.last_shot = now
            self.heat = min(self.max_heat, self.heat + self.heat_increase)
            
            # Calculate shooting direction (3 o'clock direction)
            right_angle = self.angle - 90
            direction_x = -math.sin(math.radians(right_angle))
            direction_y = -math.cos(math.radians(right_angle))
            
            # Calculate spawn position at the right side of the head
            offset_x = direction_x * (self.rect.width * 0.5)
            offset_y = direction_y * (self.rect.height * 0.5)
            
            pos_x = self.position.x + offset_x
            pos_y = self.position.y + offset_y
            
            return SoundWave(pos_x, pos_y, direction_x, direction_y)
        return None
        
    def draw_glow(self):
        from constants import screen
        # Calculate bark intensity parameters based on heat
        heat_percent = self.heat / self.max_heat
        glow_size = int(self.rect.width * (1 + heat_percent * 0.8))
        
        # Choose color based on bark intensity level
        color_index = min(len(self.glow_colors) - 1, int(heat_percent * len(self.glow_colors)))
        color = self.glow_colors[color_index]
        
        # Create bark visualization surface
        glow_surface = pygame.Surface((glow_size * 2, glow_size * 2), pygame.SRCALPHA)
        
        # Draw multiple glow circles for intensity effect
        for radius in range(glow_size, glow_size - 10, -1):
            alpha = int(100 * heat_percent * (radius / glow_size))
            glow_color = (*color, alpha)
            pygame.draw.circle(glow_surface, glow_color, (glow_size, glow_size), radius, 2)
        
        # Draw the glow on the screen
        glow_rect = glow_surface.get_rect(center=self.rect.center)
        screen.blit(glow_surface, glow_rect.topleft)
    
    def draw_shield(self):
        from constants import screen
        if not self.shield_active:
            return
            
        shield_percent = self.shield_strength / self.max_shield
        shield_size = int(self.rect.width * (1.5 + shield_percent))
        
        # Choose color based on shield strength
        color_index = min(len(self.shield_colors) - 1, int(shield_percent * len(self.shield_colors)))
        color = self.shield_colors[color_index]
        
        # Create shield surface
        shield_surf = pygame.Surface((shield_size * 2, shield_size * 2), pygame.SRCALPHA)
        
        # Draw shield as a circle with pulsating effect
        pulse = math.sin(pygame.time.get_ticks() * 0.01) * 0.1 + 0.9
        for radius in range(shield_size, shield_size - 5, -1):
            adjusted_radius = int(radius * pulse)
            alpha = int(150 * shield_percent)
            shield_color = (*color, alpha)
            pygame.draw.circle(shield_surf, shield_color, (shield_size, shield_size), adjusted_radius, 2)
        
        # Draw the shield on the screen
        shield_rect = shield_surf.get_rect(center=self.rect.center)
        screen.blit(shield_surf, shield_rect.topleft)
    
    def hide(self):
        self.hidden = True
        self.hide_timer = pygame.time.get_ticks()
        self.rect.center = (WIDTH + 200, HEIGHT + 200)
    
    def respawn(self):
        """Respawn the player in the center with invulnerability"""
        if self.lives > 0:  # Only respawn if still have lives
            self.hidden = False
            self.position = pygame.math.Vector2(WIDTH // 2, HEIGHT // 2)
            self.rect.center = self.position
            self.velocity = pygame.math.Vector2(0, 0)
            self.angle = 0
            # Make invulnerable for 2 seconds after respawn
            self.invulnerable = True
            self.invulnerable_timer = pygame.time.get_ticks()
            # Reset heat and other stats
            self.heat = 0

class Asteroid(pygame.sprite.Sprite):
    def __init__(self, size=3):
        pygame.sprite.Sprite.__init__(self)
        self.size = size
        self.image = asteroid_images[3 - self.size]
        scale = self.size * 0.3
        self.image = pygame.transform.scale(self.image, 
                                           (int(self.image.get_width() * scale),
                                            int(self.image.get_height() * scale)))
        self.rect = self.image.get_rect()
        
        # Spawn at edge of screen
        side = random.randint(1, 4)
        if side == 1:  # top
            self.rect.x = random.randint(0, WIDTH)
            self.rect.y = -self.rect.height
        elif side == 2:  # right
            self.rect.x = WIDTH
            self.rect.y = random.randint(0, HEIGHT)
        elif side == 3:  # bottom
            self.rect.x = random.randint(0, WIDTH)
            self.rect.y = HEIGHT
        else:  # left
            self.rect.x = -self.rect.width
            self.rect.y = random.randint(0, HEIGHT)
            
        self.position = pygame.math.Vector2(self.rect.center)
        
        # Random velocity
        speed = random.uniform(0.5, 2.0) * (4 - self.size)
        angle = random.uniform(0, 2 * math.pi)
        self.velocity = pygame.math.Vector2(math.cos(angle) * speed, math.sin(angle) * speed)
        
        # Random rotation
        self.rotation = 0
        self.rotation_speed = random.uniform(-3, 3)
        self.original_image = self.image
        self.trail_timer = 0
    
    def update(self):
        # Rotate asteroid
        self.rotation = (self.rotation + self.rotation_speed) % 360
        self.image = pygame.transform.rotate(self.original_image, self.rotation)
        self.rect = self.image.get_rect(center=self.rect.center)
        
        # Update position
        self.position += self.velocity
        
        # Wrap around screen
        if self.position.x < -50:
            self.position.x = WIDTH + 50
        elif self.position.x > WIDTH + 50:
            self.position.x = -50
        if self.position.y < -50:
            self.position.y = HEIGHT + 50
        elif self.position.y > HEIGHT + 50:
            self.position.y = -50
        
        self.rect.center = self.position
    
    def split(self):
        """Split asteroid into smaller pieces"""
        new_asteroids = []
        if self.size > 1:
            for _ in range(2):
                new_asteroid = Asteroid(self.size - 1)
                # Random velocity for new pieces
                angle = random.uniform(0, 2 * math.pi)
                speed = random.uniform(1, 3)
                new_asteroid.velocity = pygame.math.Vector2(
                    math.cos(angle) * speed, math.sin(angle) * speed
                )
                new_asteroids.append(new_asteroid)
        return new_asteroids

class SoundWave(pygame.sprite.Sprite):
    def __init__(self, x, y, dx, dy):
        pygame.sprite.Sprite.__init__(self)
        self.image = bullet_img
        self.rect = self.image.get_rect()
        self.rect.center = (x, y)
        self.position = pygame.math.Vector2(self.rect.center)
        self.velocity = pygame.math.Vector2(dx, dy) * 15
        self.spawn_time = pygame.time.get_ticks()
        self.lifetime = 3000
        self.original_image = self.image
        self.angle = math.degrees(math.atan2(-dy, dx)) - 90
        self.image = pygame.transform.rotate(self.original_image, self.angle)
        self.rect = self.image.get_rect(center=self.rect.center)
        self.pulse_timer = 0
        self.pulse_rate = 100
        self.trail_timer = 0
    
    def update(self):
        # Update position
        self.position += self.velocity
        self.rect.center = self.position
        
        # Pulse effect
        now = pygame.time.get_ticks()
        if now - self.pulse_timer > self.pulse_rate:
            self.pulse_timer = now
            # Scale bullet for sound wave effect
            if self.lifetime > 2000:
                scale = 1.2
                self.image = pygame.transform.scale(self.original_image, 
                                                  (int(self.original_image.get_width() * scale),
                                                   int(self.original_image.get_height() * scale)))
            else:
                self.image = self.original_image
            # Maintain rotation
            self.image = pygame.transform.rotate(self.image, self.angle)
            self.rect = self.image.get_rect(center=self.position)
        
        # Check if sound wave is off screen or expired
        if (now - self.spawn_time > self.lifetime or
            self.rect.right < 0 or self.rect.left > WIDTH or
            self.rect.bottom < 0 or self.rect.top > HEIGHT):
            self.kill()

class FireworkParticle(pygame.sprite.Sprite):
    def __init__(self, pos, velocity, color):
        pygame.sprite.Sprite.__init__(self)
        self.size = 3
        self.image = pygame.Surface((self.size, self.size), pygame.SRCALPHA)
        pygame.draw.circle(self.image, color, (self.size//2, self.size//2), self.size//2)
        self.rect = self.image.get_rect()
        self.rect.center = pos
        self.position = pygame.math.Vector2(pos)
        self.velocity = velocity
        self.gravity = pygame.math.Vector2(0, 0.05)
        self.lifetime = random.randint(30, 60)
        self.alpha = 255
        self.fade_rate = 255 / self.lifetime
        
    def update(self):
        self.velocity += self.gravity
        self.position += self.velocity
        self.rect.center = self.position
        self.lifetime -= 1
        if self.lifetime <= 0:
            self.kill()
        else:
            self.alpha = max(0, self.alpha - self.fade_rate)
            self.image.set_alpha(int(self.alpha))

File: web/sprites.js
// Sprite classes for Sheeraroids

class Sprite {
    constructor(x, y) {
        this.position = { x, y };
        this.velocity = { x: 0, y: 0 };
        this.rotation = 0;
        this.rotationSpeed = 0;
        this.size = { width: 0, height: 0 };
        this.active = true;
    }

    update() {
        // Update position based on velocity
        this.position.x += this.velocity.x;
        this.position.y += this.velocity.y;
        
        // Update rotation
        this.rotation += this.rotationSpeed;
        
        // Screen wrapping
        if (this.position.x < 0) this.position.x = WIDTH;
        if (this.position.x > WIDTH) this.position.x = 0;
        if (this.position.y < 0) this.position.y = HEIGHT;
        if (this.position.y > HEIGHT) this.position.y = 0;
    }

    draw() {
        // Override in subclasses
    }

    collidesWith(other) {
        // Simple circle collision
        const dx = this.position.x - other.position.x;
        const dy = this.position.y - other.position.y;
        const distance = Math.sqrt(dx * dx + dy * dy);
        return distance < (this.radius + other.radius);
    }
}

class Sheera extends Sprite {
    constructor(gameMode) {
        super(WIDTH / 2, HEIGHT / 2);
        this.radius = 20;
        this.rotation = -Math.PI / 2; // Point upward
        this.rotationSpeed = 0;
        this.acceleration = 0.2;
        this.friction = 0.98;
        this.maxSpeed = 5;
        this.lives = 3;
        this.invulnerable = false;
        this.invulnerableTime = 0;
        this.hidden = false;
        this.shootCooldown = 0;
        this.heat = 0;
        this.maxHeat = 100;
        this.heatCooldownRate = 0.5;
        this.shield = false;
        this.shieldStrength = 100;
        this.maxShield = 100;
        this.shieldRechargeRate = 0.1;
        this.shieldActive = false;
        this.shieldColors = [
            '#00FFFF', '#00CCFF', '#0099FF', '#0066FF', '#0033FF'
        ];
        
        // Apply game mode modifiers
        if (gameMode === "accelerated") {
            this.acceleration *= 1.5;
            this.maxSpeed *= 1.5;
            this.rotationSpeed *= 1.5;
        } else if (gameMode === "slowed") {
            this.acceleration *= 0.5;
            this.maxSpeed *= 0.5;
            this.rotationSpeed *= 0.5;
        }
        
        // Load ship image
        this.image = new Image();
        this.image.src = ASSETS_PATH + 'ship.png';
    }

    update() {
        if (this.hidden) return;
        
        super.update();
        
        // Apply friction
        this.velocity.x *= this.friction;
        this.velocity.y *= this.friction;
        
        // Cool down weapon heat
        if (this.heat > 0) {
            this.heat -= this.heatCooldownRate;
            if (this.heat < 0) this.heat = 0;
        }
        
        // Recharge shield
        if (!this.shieldActive && this.shieldStrength < this.maxShield) {
            this.shieldStrength += this.shieldRechargeRate;
            if (this.shieldStrength > this.maxShield) {
                this.shieldStrength = this.maxShield;
            }
        }
        
        // Handle invulnerability timer
        if (this.invulnerable) {
            this.invulnerableTime--;
            if (this.invulnerableTime <= 0) {
                this.invulnerable = false;
            }
        }
        
        // Decrease shoot cooldown
        if (this.shootCooldown > 0) {
            this.shootCooldown--;
        }
    }

    draw() {
        if (this.hidden) return;
        
        ctx.save();
        ctx.translate(this.position.x, this.position.y);
        ctx.rotate(this.rotation);
        
        // Draw ship
        ctx.drawImage(this.image, -this.radius, -this.radius, this.radius * 2, this.radius * 2);
        
        // Draw shield if active
        if (this.shieldActive) {
            this.drawShield();
        }
        
        // Draw heat glow if hot
        if (this.heat > 0) {
            this.drawGlow();
        }
        
        ctx.restore();
    }

    drawShield() {
        const shieldRadius = this.radius * (1.5 + this.shieldStrength / this.maxShield * 0.5);
        const colorIndex = Math.min(this.shieldColors.length - 1, 
                                   Math.floor((this.shieldStrength / this.maxShield) * this.shieldColors.length));
        
        ctx.beginPath();
        ctx.arc(0, 0, shieldRadius, 0, Math.PI * 2);
        ctx.strokeStyle = this.shieldColors[colorIndex];
        ctx.lineWidth = 3;
        ctx.globalAlpha = 0.7;
        ctx.stroke();
        ctx.globalAlpha = 1.0;
    }

    drawGlow() {
        const glowRadius = this.radius * (1 + this.heat / this.maxHeat * 0.5);
        const gradient = ctx.createRadialGradient(0, 0, this.radius, 0, 0, glowRadius);
        
        gradient.addColorStop(0, 'rgba(255, 100, 0, 0.8)');
        gradient.addColorStop(1, 'rgba(255, 50, 0, 0)');
        
        ctx.beginPath();
        ctx.arc(0, 0, glowRadius, 0, Math.PI * 2);
        ctx.fillStyle = gradient;
        ctx.fill();
    }

    accelerate() {
        const dx = Math.cos(this.rotation) * this.acceleration;
        const dy = Math.sin(this.rotation) * this.acceleration;
        
        this.velocity.x += dx;
        this.velocity.y += dy;
        
        // Limit speed
        const speed = Math.sqrt(this.velocity.x * this.velocity.x + this.velocity.y * this.velocity.y);
        if (speed > this.maxSpeed) {
            const ratio = this.maxSpeed / speed;
            this.velocity.x *= ratio;
            this.velocity.y *= ratio;
        }
    }

    rotate(direction) {
        this.rotationSpeed = direction * 0.05;
    }

    stopRotation() {
        this.rotationSpeed = 0;
    }

    shoot() {
        if (this.shootCooldown > 0 || this.hidden) return null;
        
        // Add heat when shooting
        this.heat += 10;
        if (this.heat > this.maxHeat) {
            this.heat = this.maxHeat;
            return null; // Too hot to shoot
        }
        
        this.shootCooldown = 10;
        
        return new SoundWave(
            this.position.x + Math.cos(this.rotation) * this.radius,
            this.position.y + Math.sin(this.rotation) * this.radius,
            this.rotation
        );
    }

    activateShield() {
        if (this.shieldStrength > 0 && !this.hidden) {
            this.shieldActive = true;
        }
    }

    deactivateShield() {
        this.shieldActive = false;
    }

    hide() {
        this.hidden = true;
        this.invulnerable = true;
        this.invulnerableTime = 180; // 3 seconds at 60fps
        
        // Respawn after 2 seconds
        setTimeout(() => {
            this.position.x = WIDTH / 2;
            this.position.y = HEIGHT / 2;
            this.velocity.x = 0;
            this.velocity.y = 0;
            this.hidden = false;
        }, 2000);
    }
}

class SoundWave extends Sprite {
    constructor(x, y, angle) {
        super(x, y);
        this.radius = 5;
        this.speed = 10;
        this.velocity.x = Math.cos(angle) * this.speed;
        this.velocity.y = Math.sin(angle) * this.speed;
        this.spawnTime = Date.now();
        this.lifetime = 2000; // 2 seconds
    }

    update() {
        super.update();
        
        // Check if bullet has expired
        if (Date.now() - this.spawnTime > this.lifetime) {
            this.active = false;
        }
    }

    draw() {
        ctx.save();
        ctx.translate(this.position.x, this.position.y);
        
        // Draw sound wave
        ctx.beginPath();
        ctx.arc(0, 0, this.radius, 0, Math.PI * 2);
        ctx.fillStyle = '#00FFFF';
        ctx.fill();
        
        // Draw wave rings
        const age = (Date.now() - this.spawnTime) / 100;
        for (let i = 0; i < 3; i++) {
            const ringRadius = this.radius + (i + 1) * 3 + Math.sin(age + i) * 2;
            ctx.beginPath();
            ctx.arc(0, 0, ringRadius, 0, Math.PI * 2);
            ctx.strokeStyle = `rgba(0, 255, 255, ${0.7 - i * 0.2})`;
            ctx.lineWidth = 1;
            ctx.stroke();
        }
        
        ctx.restore();
    }
}

class Asteroid extends Sprite {
    constructor(size, x, y) {
        // If position not provided, use random position
        x = x || Math.random() * WIDTH;
        y = y || Math.random() * HEIGHT;
        
        super(x, y);
        
        this.size = size || 3; // 3 = large, 2 = medium, 1 = small
        this.radius = this.size * 20;
        
        // Random velocity
        const speed = (4 - this.size) * 0.5 + 0.5;
        const angle = Math.random() * Math.PI * 2;
        this.velocity.x = Math.cos(angle) * speed;
        this.velocity.y = Math.sin(angle) * speed;
        
        // Random rotation
        this.rotationSpeed = (Math.random() - 0.5) * 0.04;
        
        // Create asteroid shape (vertices)
        this.vertices = [];
        const numVertices = 10 + this.size * 2;
        for (let i = 0; i < numVertices; i++) {
            const angle = (i / numVertices) * Math.PI * 2;
            const distance = this.radius * (0.8 + Math.random() * 0.4);
            this.vertices.push({
                x: Math.cos(angle) * distance,
                y: Math.sin(angle) * distance
            });
        }
        
        // Load iguana image
        this.image = new Image();
        this.image.src = ASSETS_PATH + 'iguana.png';
    }

    draw() {
        ctx.save();
        ctx.translate(this.position.x, this.position.y);
        ctx.rotate(this.rotation);
        
        // Draw asteroid (iguana)
        ctx.drawImage(this.image, -this.radius, -this.radius, this.radius * 2, this.radius * 2);
        
        ctx.restore();
    }

    split() {
        if (this.size <= 1) {
            return []; // Too small to split
        }
        
        const newAsteroids = [];
        for (let i = 0; i < 2; i++) {
            const newAsteroid = new Asteroid(this.size - 1, this.position.x, this.position.y);
            newAsteroids.push(newAsteroid);
        }
        
        return newAsteroids;
    }
}

class FireworkParticle extends Sprite {
    constructor(x, y, velocity, color) {
        super(x, y);
        this.velocity = velocity || { x: (Math.random() - 0.5) * 5, y: (Math.random() - 0.5) * 5 };
        this.color = color || '#FFFFFF';
        this.radius = 2;
        this.lifetime = 60; // 1 second at 60fps
        this.initialLifetime = this.lifetime;
    }

    update() {
        super.update();
        
        // Apply gravity
        this.velocity.y += 0.1;
        
        // Decrease lifetime
        this.lifetime--;
        if (this.lifetime <= 0) {
            this.active = false;
        }
    }

    draw() {
        const alpha = this.lifetime / this.initialLifetime;
        
        ctx.save();
        ctx.globalAlpha = alpha;
        ctx.beginPath();
        ctx.arc(this.position.x, this.position.y, this.radius, 0, Math.PI * 2);
        ctx.fillStyle = this.color;
        ctx.fill();
        ctx.restore();
    }
}